File: frontend/app/server/routes/projects.ts
import { selectCurrentBranch, selectCurrentBranchId } from "@/app/store/features/workspace/selectors/commitHistory";
import { resetDeltasForBranch } from "@/app/store/features/workspace/workspaceSlice";
import store from "@/app/store/store";
import { Id } from "@/app/types/common";
import _ from "lodash";
import { Project } from "next/dist/build/swc";
import { sendAndHandle } from "../handle";
import { USER_ID } from '../constants';
import Root from "@/app/proto/compiled"
import { isDeltasEmpty } from "../helpers";
import { BackwardDeltas, Branch, Commit, Graph } from "@/app/types/main";









export function emitCreateProject(project: Project) {
    const projectCopy = _.cloneDeep(project) as any
    delete projectCopy["branchOrder"]
    delete projectCopy["graphs"]
    if (projectCopy.branches) {
      Object.values(projectCopy.branches).forEach(branch => {
        delete branch.backwardDeltas;
        delete branch.forwardDeltas;
      });
    } 
    projectCopy.ownerId = USER_ID
    
    const projectPB = Root.Project.create(projectCopy)
    const messageWrapper = Root.Wrapper.create()
    messageWrapper.projectPB = projectPB

    const messages = Root.MessageList.create()
    messages.list.push(messageWrapper)


    sendAndHandle("projects/create", "POST", [["user_id", USER_ID]],  messages )
}

export function emitProjectDelete(id: Id) {
    sendAndHandle("projects/delete", "POST", [["project_id", id]])
}

export function emitCreateCommit(projectId: Id, branchId: Id, commitId: Id, commit: Commit, deltas: BackwardDeltas) {


    const commitMessage = Root.Commit.create(commit)
    const backwardDeltasMessage = Root.BackwardDeltas.create(deltas)
    
    
    const messageWrapper1 = Root.Wrapper.create()
    messageWrapper1.commitPB = commitMessage

    const messageWrapper2 = Root.Wrapper.create()
    messageWrapper2.backwardDeltasPB = backwardDeltasMessage

    const messages = Root.MessageList.create()
    messages.list.push(messageWrapper1)
    messages.list.push(messageWrapper2)


    sendAndHandle(
        "projects/create-commit", 
        "POST", 
        [["user_id", USER_ID], ["project_id", projectId], ["branch_id", branchId], ["commit_id", commitId]],
        messages
    )
}

export function emitCreateBranch(projectId: Id, branchId: Id, branch: Branch, currentSave: Graph) {

    const branchMessage = Root.Branch.create(branch)
    const graphMessage = Root.Graph.create(currentSave)
    
    const messageWrapper1 = Root.Wrapper.create()
    messageWrapper1.branchPB = branchMessage
    
    const messageWrapper2 = Root.Wrapper.create()
    messageWrapper2.graphPB = graphMessage
    
    const messages = Root.MessageList.create()
    messages.list.push(messageWrapper1)
    messages.list.push(messageWrapper2)

    sendAndHandle(
        "projects/create-branch", 
        "POST", 
        [["user_id", USER_ID], ["project_id", projectId], ["branch_id", branchId]],
        messages
    )
}

export function emitDeleteBranch(projectId: Id, branchId: Id) {
    sendAndHandle(
        "projects/delete-branch", 
        "POST", 
        [["user_id", USER_ID], ["project_id", projectId], ["branch_id", branchId]]
    )
}

export function emitUpdateBranchName(projectId: Id, branchId: Id, newName: string) {
    sendAndHandle(
        "projects/update-branch-name", 
        "POST", 
        [ ["project_id", projectId], ["branch_id", branchId],  ['new_name', newName]]
    )
}


export function emitDeltas() {

    //Foreward deltas are used by the server to update its graph
    //Backward deltas are used by the client to calculate commit graphs.
    //Backward deltas are not modified in any way by the server. They are immediately saved to its file system.

    const state = store.getState()
    const dispatch = store.dispatch



    const branch = selectCurrentBranch(state.workspace)
    const branchId = selectCurrentBranchId(state.workspace)
    const forwardDeltas = branch?.forwardDeltas
    const backwardDeltas = branch?.backwardDeltas

    if (!branchId || !backwardDeltas || !forwardDeltas ||  isDeltasEmpty(forwardDeltas)) return
    const forwardDeltasMessage = Root.ForwardDeltas.create(forwardDeltas)
    const backwardDeltasMessage = Root.BackwardDeltas.create(backwardDeltas)

    const message1 = Root.Wrapper.create()
    message1.forwardDeltasPB = forwardDeltasMessage as Root.ForwardDeltas
    const message2 = Root.Wrapper.create()
    message2.backwardDeltasPB = backwardDeltasMessage as Root.BackwardDeltas

    const messages = Root.MessageList.create()
    messages.list.push(message1)
    messages.list.push(message2)

    
    sendAndHandle(
        "projects/graph-deltas", 
        "POST", 
        [["user_id", USER_ID], ["branch_id", branchId] ],
        messages
    ).then(() => {
        dispatch(resetDeltasForBranch({ name: branch.name, deltaType: 'forwardDeltas'}))
    })
}

File: computeserver/server/app/routers/projects.py
from fastapi import APIRouter, File, Query
from typing import Annotated

from computeserver.server.protos import wrapper_pb2

from computeserver.server.services.file_utils.graph import apply_deltas, create_graph, delete_graph, write_graph_to_file
from computeserver.server.services.file_utils.deltas import write_deltas_to_file
from computeserver.server.services.file_utils.safe_tensors import create_safe_tensor_file
import computeserver.server.database.database_operations as db



router = APIRouter()


router = APIRouter(
    prefix="/projects",
    tags=["projects"],
)




@router.post("/create")
async def create_project(user_id: Annotated[str, Query()], file: Annotated[bytes, File()]):

    messages = wrapper_pb2.MessageList()
    messages.ParseFromString(file)
    project_pb = messages.list[0].projectPB

    main_branch_id = list(project_pb.branches.keys())[0]
    create_graph(user_id, main_branch_id)
    create_safe_tensor_file(user_id, main_branch_id)
    db.create_project(project_pb)

    return { "user_id": user_id }


@router.post('/delete')
async def delete_project(project_id: str):
    db.delete_project(project_id)
    return { "project_id": project_id }


@router.post('/create-commit')
async def create_commit(user_id: str, project_id: str, branch_id: str, commit_id: str, file: Annotated[bytes, File()]):

    messages = wrapper_pb2.MessageList()
    messages.ParseFromString(file)

    commit_pb = messages.list[0].commitPB
    backward_deltas_pb = messages.list[1].backwardDeltasPB

    latest_commit_id = db.get_latest_commit_id(project_id, branch_id)
    if latest_commit_id:
        db.append_id_to_commit_next(project_id, latest_commit_id, commit_id)
    db.set_branch_latest_commit_id(project_id, branch_id, commit_id)
    write_deltas_to_file(user_id, commit_id, backward_deltas_pb.SerializeToString())
    db.insert_commit(project_id, commit_pb, commit_id)

    return {"user_id": user_id, "branch_id": branch_id, "commit_id": commit_id}


@router.post("/create-branch")
async def create_branch(user_id: str, project_id: str, branch_id: str, file: Annotated[bytes, File()] ):

    messages = wrapper_pb2.MessageList()
    messages.ParseFromString(file)

    branch_pb = messages.list[0].branchPB
    graph_pb = messages.list[1].graphPB

    write_graph_to_file(user_id, branch_id, graph_pb)

    db.insert_branch(project_id,branch_pb, branch_id)

    return {"user_id": user_id, "branch_id": branch_id}


@router.post("/delete-branch")
async def delete_branch(user_id: str, project_id: str, branch_id: str):
    db.delete_branch(project_id, branch_id)
    delete_graph(user_id, branch_id)

    return {"user_id": user_id, "branch_id": branch_id}



@router.post("/update-branch-name")
async def update_branch_name(project_id: str, branch_id: str, new_name: str):
    db.update_branch_name(project_id, branch_id, new_name)

    return { "branch_id": branch_id, "new_name": new_name }



@router.post("/graph-deltas")
async def graph_deltas(user_id: str, branch_id: str, file: Annotated[bytes, File()]):
    
    messages = wrapper_pb2.MessageList()
    messages.ParseFromString(file)

    forward_deltas = messages.list[0].forwardDeltasPB
    backward_deltas  = messages.list[1].backwardDeltasPB

    apply_deltas(user_id, branch_id, forward_deltas)

    write_deltas_to_file(user_id, branch_id, backward_deltas.SerializeToString())

    return {"user_id": user_id, "branch_id": branch_id}